### PR TITLE
8340141: C1: rework ciMethod::equals following 8338471

### PR DESCRIPTION
**Pull request by dean-long**
This PR changes ciMethod::equals() to a special-purpose debug helper method for the one place in C1 that uses it in an assert.  The reason why making it general purpose is difficult is because JVMTI can add and delete methods.   See the bug report and JDK-8338471 for more details.  I'm open to suggestions for a better name than equals_ignore_version().

An alternative approach, which I think may actually be better, would be to check for old methods first, and bail out if we see any.  Then we can change the assert back to how it was originally, using ==.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340141](https://bugs.openjdk.org/browse/JDK-8340141): C1: rework ciMethod::equals following 8338471 (**Bug** - P4)

### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21148/head:pull/21148` \
`$ git checkout pull/21148`

Update a local copy of the PR: \
`$ git checkout pull/21148` \
`$ git pull https://git.openjdk.org/jdk.git pull/21148/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21148`

View PR using the GUI difftool: \
`$ git pr show -t 21148`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21148.diff">https://git.openjdk.org/jdk/pull/21148.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21148#issuecomment-2369869920)
</details>

### CODE REVIEW

**Comment by dean-long** [comment]
The alternative approach would look something like this:
```
    if (target->get_Method()->is_old() || cha_monomorphic_target->get_Method()->is_old()) {
        BAILOUT("redefined method");
    }
    assert(!target->can_be_statically_bound() || target == cha_monomorphic_target, "");
```

**Comment by dean-long** [comment]
@coleenp, @matias9927, please take a look at this PR too.

**Comment by vnkozlov** [comment]
Instead of bailout in alternative approach we can change `cha_monomorphic_target` to `nullptr` in code which is looking for it in previous lines. `target` will be used for call  and we will loose a little performance when JVMTI is used instead of skipping compilation. Am I missing something?

**Comment by dean-long** [comment]
@vnkozlov , I like the alternative approach better.  I went with the current approach because I was thinking it would be simpler than the bailout, but I changed my mind after writing both out.

Yes, we can change cha_monomorphic_target to nullptr instead of bailing out.  But my understanding is any use of old/redefined methods will cause the compilation to be thrown out when we try to create the nmethod, so we are avoiding wasted work by bailing out early.

**Comment by vnkozlov** [comment]
>Yes, we can change cha_monomorphic_target to nullptr instead of bailing out. But my understanding is any use of old/redefined methods will cause the compilation to be thrown out when we try to create the nmethod, so we are avoiding wasted work by bailing out early.

Good point.

**Comment by vnkozlov** [comment]
@dean-long, may be I misunderstand your statement. Are you re-writing the fix or keep current?

**Comment by iwanowww** [comment]
> JVMTI can add and delete methods

Can you elaborate on that point, please? JVMTI spec states that redefinition/retransformation "must not add, remove or rename fields or methods" [1] [2].

[1] https://docs.oracle.com/en/java/javase/23/docs/specs/jvmti.html#RedefineClasses
[2] https://docs.oracle.com/en/java/javase/23/docs/specs/jvmti.html#RetransformClasses

**Comment by iwanowww** [comment]
I like @vnkozlov suggestion to null out `cha_monomorphic_target`. Moreover, the validation can be performed inside `ciMethod::find_monomorphic_target()` which is used to compute `cha_monomorphic_target`.

File: src/hotspot/share/ci/ciMethod.cpp [hunk 1]
```diff
@@ -781,20 +781,33 @@ bool ciMethod::can_omit_stack_trace() const {
...
+// Usage note: this is not a general-purpose API, but is only used for a C1
+// assert comparing two non-private methods, which means they cannot be
+// deleted.  To make this a general-purpose API, the compilers would need to
+// deal gracefully with added/deleted methods.
+bool ciMethod::equals_ignore_version(const ciMethod* m) const {
   if (this == m) return true;
   VM_ENTRY_MARK;
   Method* m1 = this->get_Method();
   Method* m2 = m->get_Method();
+  guarantee(!m1->is_private() && !m1->is_deleted(), "see usage note");
```
Comment (iwanowww) [inline src/hotspot/share/ci/ciMethod.cpp:800]: Some changes inside `ciMethod::equals` look irrelevant to checking method equality  (e.g., asserting that a method is not private). Alternatively, if you decide to keep the current shape of the fix, the code can be moved closer to the use site as a helper function.

**Comment by dean-long** [comment]
> @dean-long, may be I misunderstand your statement. Are you re-writing the fix or keep current?

Either one is fine with me.  I could make a separate draft PR with the alternative solution if that helps reviewers decide.

**Comment by dean-long** [comment]
> > JVMTI can add and delete methods
> 
> Can you elaborate on that point, please? JVMTI spec states that redefinition/retransformation "must not add, remove or rename fields or methods" [1] [2].
> 
> [1] https://docs.oracle.com/en/java/javase/23/docs/specs/jvmti.html#RedefineClasses [2] https://docs.oracle.com/en/java/javase/23/docs/specs/jvmti.html#RetransformClasses

It's because of the AllowRedefinitionToAddDeleteMethods flag:

https://github.com/openjdk/jdk/blob/7312eea382eed048b6abdb6409c006fc8e8f45b4/src/hotspot/share/prims/jvmtiRedefineClasses.cpp#L928

**Comment by dean-long** [comment]
> I like @vnkozlov suggestion to null out `cha_monomorphic_target`. Moreover, the validation can be performed inside `ciMethod::find_monomorphic_target()` which is used to compute `cha_monomorphic_target`.

I like this idea.  I pushed a new version.

**Review by vnkozlov — APPROVED** [review]
This looks good.

**Comment by dean-long** [comment]
Thanks @vnkozlov.

File: src/hotspot/share/ci/ciMethod.cpp [hunk 2]
```diff
@@ -688,6 +688,11 @@ ciMethod* ciMethod::find_monomorphic_target(ciInstanceKlass* caller,
     return nullptr;
   }
 
+  // Redefinition support.
+  if (this->get_Method()->is_old() || root_m->get_Method()->is_old()) {
```
Comment (iwanowww) [inline src/hotspot/share/ci/ciMethod.cpp:692]: Is it safe to access raw `Method*` from a compiler thread which is not in VM state?

Comment (dean-long) [reply src/hotspot/share/ci/ciMethod.cpp:692]: No, probably not.  I'll fix it.  I was assuming the whole function was in the VM state.

File: src/hotspot/share/ci/ciMethod.cpp [hunk 3]
```diff
@@ -682,12 +682,19 @@ ciMethod* ciMethod::find_monomorphic_target(ciInstanceKlass* caller,
...
+
   ciMethod* root_m = resolve_invoke(caller, actual_recv, check_access, true /* allow_abstract */);
   if (root_m == nullptr) {
     // Something went wrong looking up the actual receiver method.
     return nullptr;
   }
 
+  // Redefinition support.
+  if (this->is_old() || root_m->is_old()) {
+    return nullptr;
```
Comment (iwanowww) [inline src/hotspot/share/ci/ciMethod.cpp:695]: IMO you can safely drop this particular check. The one after `Dependencies::find_unique_concrete_method()` should be enough to preserve the invariant (`target == cha_monomorphic_target`) .

But thinking more about it, now I'm curious what happens when an old method is actually encountered. The fix conservatively rejects possible inlining opportunity, but it seems it doesn't invalidate resulting nmethod anymore. So, no recompilation attempt follows to recuperate that.

We could either record a evol dependency on a stale `Method` (to fail during nmethod installation step) or fail-fast the compilation (probably, implies additional checks to propagate the failure status).

Comment (dean-long) [reply src/hotspot/share/ci/ciMethod.cpp:695]: Even if we check for stale Methods in various places, including invoke(), there is nothing to prevent the method from going stale after the last spot-check.  My understanding was that we already handle stale metadata as a precondition to creating the nmethod.  If we have a loophole there that lets stale metadata get through, then that's a separate existing bug for C1 and C2.
I was tempted to add a bailout, but the reason would be as a performance improvement to short-circuit wasted work, not to correct a stale metadata problem.

Comment (dean-long) [reply src/hotspot/share/ci/ciMethod.cpp:695]: > IMO you can safely drop this particular check. The one after Dependencies::find_unique_concrete_method() should be enough to preserve the invariant (target == cha_monomorphic_target) .

If I do that, then I can also revert the VM state changes.  However, I wasn't able to convince myself that this check is not needed.  If we end up returning root_m here as cha_monomorphic_target, it seems possible that it could be a new version of the method, and then target == cha_monomorphic_target would fail.

Comment (dean-long) [reply src/hotspot/share/ci/ciMethod.cpp:695]: C1 does call 
```
2159            dependency_recorder()->assert_evol_method(inline_target);
```
which in the CHA case would be `cha_monomorphic_target`, not `target`, so it looks like we may not detect if `target` is stale as long as `cha_monomorphic_target` is not.  It seems like a minor loophole, but I'm not sure what kind of problems it could cause, especially if the bytecodes of `target` are not used.

**Comment by dean-long** [comment]
Hold off on re-reviews.  I need to fix some errors introduced by moving the VM state transitions.

**Comment by dean-long** [comment]
OK, fixed version pushed.  I moved the first group of is_old checks into resolve_invoke().

File: src/hotspot/share/ci/ciMethod.cpp (hunk 3 above)
Comment (iwanowww) [reply src/hotspot/share/ci/ciMethod.cpp]: The scenario which concerns me is performance-related. If CHA conservatively disables inlining when concurrent class redefinition takes place during parsing, then there are no mechanism in place to recuperate possible loss of performance. Normally, if inlined method is redefined later during compilation, nmethod installation fails during dependency validation. But here no inlining happens (missed optimization opportunity) and call sites in generated code are resolved based on symbolic information (except rare cases when resolved method is attached to the call site, see `SharedRuntime::find_callee_info_helper()` for details), so there are no guarantees the stale `Method*` is recorded.

I agree that the window for such sequence of events is narrow, but it may be a source of surprising performance anomalies in rare cases.

Comment (dean-long) [reply src/hotspot/share/ci/ciMethod.cpp]: OK, it sounds like we have two choices.  Either record an evol dependency every time we short-circuit an optimization based on is_old, or bail out.  I vote for bailing out.

Comment (vnkozlov) [reply src/hotspot/share/ci/ciMethod.cpp]: I vote for bail out.

**Review by vnkozlov — APPROVED** [review]
Good.

File: src/hotspot/share/ci/ciMethod.cpp (hunk 3 above)
Comment (iwanowww) [reply src/hotspot/share/ci/ciMethod.cpp]: I prefer bailing out as well, but, please, check it doesn't mark the root method as non-compilable.

**Comment by dean-long** [comment]
Added bailouts.  Because we record failure in CI layer, C1 and C2 need to check for failure there.  C2 already did that, but C1 did not.  For C1 I decided to delegate all failure recording to the CI layer.  This is a step towards implementing JDK-8132354, and avoids string copying issues like those in JDK-8325095.

**Comment by dean-long** [comment]
I'm still working on this.  It's surprisingly tricky to add new bailout failure points.  We also have to add new bailout check locations so we don't try to continue while in an inconsistent state (same issues as exception handling).
I'm simulating redefined methods randomly (similar to C2 StressBailout and fail_randomly()) to stress test it.

**Comment by dean-long** [comment]
New version.  I was going to add dependencies for any old methods found, but then I realized we check jvmti_state_changed() at the end, which will throw out the compilation if any methods are redefined during the compilation.

**Comment by dean-long** [comment]
Thanks @iwanowww.
@vnkozlov , please take another look when you get a chance.

**Review by vnkozlov — APPROVED** [review]
This version looks good to me too.

**Comment by dean-long** [comment]
Thanks @vnkozlov.
